tests/unit/s2n_cipher_suite_match_test.c

```diff
@@ -781,6 +781,443 @@ int main(int argc, char **argv)
             }
         }
 
+        /* Test ChaCha20 boosting behaviour for sslv2 connection */
+        {
+            struct s2n_cipher_preferences cipher_preferences = { 0 };
+
+            struct s2n_security_policy security_policy = {
+                .minimum_protocol_version = S2N_SSLv2,
+                .cipher_preferences = &cipher_preferences,
+                .kem_preferences = &kem_preferences_null,
+                .signature_preferences = &s2n_signature_preferences_20201021,
+                .ecc_preferences = &s2n_ecc_preferences_test_all,
+            };
+    
+            DEFER_CLEANUP(struct s2n_connection *connection = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(connection);
+
+            connection->security_policy_override = &security_policy;
+            connection->kex_params.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+            connection->client_protocol_version = S2N_SSLv2;
+            connection->actual_protocol_version = S2N_SSLv2;
+            connection->server_protocol_version = S2N_SSLv2;
+
+            /* Need to set the server config for tls 1.2 ciphers so that we pass auth validation */
+            EXPECT_SUCCESS(s2n_connection_set_config(connection, server_config));
+
+            static struct s2n_cipher_suite *cipher_suites_sslv2[] = {
+                &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256,       /* 0xC0,0x23 */
+                &s2n_ecdhe_ecdsa_with_chacha20_poly1305_sha256, /* 0xCC,0xA9 */
+                &s2n_ecdhe_rsa_with_chacha20_poly1305_sha256,   /* 0xCC,0xA8 */
+                &s2n_dhe_rsa_with_chacha20_poly1305_sha256,     /* 0xCC,0xAA */
+            };
+
+            cipher_preferences = (struct s2n_cipher_preferences) {
+                .count = s2n_array_len(cipher_suites_sslv2),
+                .suites = cipher_suites_sslv2,
+                .allow_chacha20_boosting = true
+            };
+
+            
+            /* Server and client have chacha20 boosting enabled. SSLv2 Server negotiates its most preferred
+             * chacha20 cipher suite (which happens to NOT be the client's most preferred)
+             */
```

Comment:
Is this comment right? The server's most preferred chacha20 cipher suite is s2n_ecdhe_ecdsa_with_chacha20_poly1305_sha256, and the client's is TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256, and s2n_ecdhe_ecdsa_with_chacha20_poly1305_sha256 is chosen. The server has a more preferred non-chacha20 suite-- is that what you meant? I'm also not sure why the wire list is called "ecdhe_rsa_chacha20_boosted_wire", when its top chacha20 is ecdsa.

You're also still not actually testing chacha20 boosting here. Since s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256 isn't offered by the client, s2n_ecdhe_ecdsa_with_chacha20_poly1305_sha256 would be chosen even without chacha20 boosting. Since you've made this mistake a couple times, that might be a sign that you need some sanity checks in your test. Try toggling "enable_chacah20_boosting" on and off in your tests and assert that a different cipher suite is chosen.

Also: I still think it would be useful to try to exercise the case where we iterate all the way through the wire ciphers. You should be able to do that easily by making s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256 the last entry in the wire ciphers. s2n_set_cipher_as_sslv2_server is no longer constructing the stuffer used by s2n_wire_ciphers_contain, so it's not mandatory, but it is easy to do.